plugins/pulp_python/plugins/models.py

```diff
@@ -1,120 +1,150 @@
-                                            metadata, flags=re.MULTILINE).group('field')
-            except AttributeError:
-                msg = _('The PKG-INFO file is missing required attributes. Please ensure that the '
-                        'following attributes are all present: %(attrs)s')
-                msg = msg % {
-                    'attrs': ', '.join([cls._metadata_label(attr) for attr in REQUIRED_ATTRS])}
-                raise ValueError(msg)
-
-            # Add the filename, checksum, and checksum_type to the attrs
-            attrs['_filename'] = '%s-%s.tar%s' % (attrs['name'], attrs['version'], compression_type)
-            attrs['_checksum'] = checksum
-            attrs['_checksum_type'] = DEFAULT_CHECKSUM_TYPE
-            attrs['_metadata_file'] = metadata_file_name
-            package = cls(**attrs)
-            return package
-        finally:
-            if 'package_archive' in locals():
-                package_archive.close()
+    def from_json(cls, package_data, release, project_data):
+        """
+        Create and return (but do not save) an instance of a single Python package object from the
+        metadata available from PyPI JSON.
+
+        :param package_data: metadata specific to a file
+        :type  package_data: dict
+        :param release: version number
+        :type  release: basestring
+        :param project_data: metadata that applies to all versions
+        :type  project_data: dict
+        :return: instance of a package
+        :rtype:  pulp_python.plugins.models.Package
+        """
+        package_attrs = {}
+
+        package_attrs['version'] = release
+        package_attrs['name'] = project_data['name']
+        package_attrs['author'] = project_data['author']
+        package_attrs['summary'] = project_data['summary']
+
+        package_attrs['filename'] = package_data['filename']
+        package_attrs['path'] = package_data['path']
+        package_attrs['packagetype'] = package_data['packagetype']
+        package_attrs['md5_digest'] = package_data['md5_digest']
+
+        # If we are syncing from PyPI, there will be no `checksum`, but will be `md5_digest`
+        package_attrs['_checksum'] = package_data.get('checksum', package_attrs['md5_digest'])
+        package_attrs['_checksum_type'] = package_data.get('checksum_type', 'md5')
+        return cls(**package_attrs)
+
+    @classmethod
+    def from_archive(cls, path):
+        """
+        Create and return (but do not save) an instance of a Python package object from the package
+        itself.
+
+        Twine is smart enough to crack open a tarball, zip, or wheel and parse the metadata that is
+        contained in the package.
+
+        :param path: path to the package
+        :type  path: basestring
+        :return: instance of a package
+        :rtype:  pulp_python.plugins.models.Package
+        """
+        meta_dict = PackageFile.from_filename(path, comment='').metadata_dictionary()
```

Comment:
The metadata dictionary coming back from twine looks to include at least ``md5_digest`` (in the future I expect there will be ``sha256_digest`` as well). It would be good if, at this point, the file was confirmed to match the checksum in the metadata. On https://github.com/pulp/pulp_python/pull/84/files#diff-3905b2a2875b436d8a5d0e7818c6826eR146 the checksum is calculated again and it could be that this is performed on a corrupt file, right?

https://github.com/pypa/twine/blob/a37fcbd2e4631144f1de2912d4ee99bc3d5d1bb8/twine/package.py#L92